### ISSUE TITLE
perf: DiT-XL / VGG16BN test workload bottlenecks seen from downstream AiDotNet repo

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#162]
Filed from [ooples/AiDotNet#1136](https://github.com/ooples/AiDotNet/issues/1136) after running a series of perf PRs against cancelled CI shards on large model tests. Documenting what's observably rate-limiting so it can inform engine-level priorities if/when they align with what's actually shipped here.

## Context

The [AiDotNet](https://github.com/ooples/AiDotNet) library uses this package as its compute backend. Several CI shards there run production-default model sizes (VGG16BN at 224×224, DiT-XL at hidden=1152 × 28 blocks, full-vocab embedding transformers) through 30-iteration training loops and hit the 45-minute workflow wall clock. Construction-side memory issues were tackled in AiDotNet via lazy init + pool return (PRs [#1137](https://github.com/ooples/AiDotNet/pull/1137), [#1140](https://github.com/ooples/AiDotNet/pull/1140), [#1141](https://github.com/ooples/AiDotNet/pull/1141)). Runtime speed is now the remaining ceiling.

## Observation

On the AiDotNet side, every hot layer (`DenseLayer.Forward`, `ConvolutionalLayer.Forward`, `SelfAttentionLayer.Forward`, `LayerNormalizationLayer.Forward`, `MultiHeadAttentionLayer`) already routes through `Engine.X` operations. The per-layer glue is not the bottleneck. What tests observably spend their wall-clock time on:

1. `Engine.TensorMatMul` / `Engine.FusedLinear` — called from every dense-style projection, ~112 calls per DiT-XL forward pass.
2. `Engine.Conv2D` / `Engine.Conv2DBackwardKernel` — 13–16 calls per VGG forward, more for diffusion U-Nets.
3. `Engine.ScaledDotProductAttention` — 28 calls per DiT-XL forward.
4. `Engine.LayerNorm` — many per-transformer-block calls.

Specific test-level symptoms pulled from the run log on [AiDotNet PR #1135 run 24398739627](https://github.com/ooples/AiDotNet/actions/runs/24398739627):

- 94 per-test `Test execution timed out after 120000 milliseconds` on the `ModelFamily - NeuralNetworks` shard, concentrated on VGG / Capsule / SpikingNN / HTM / TransformerEmbedding / VariationalAutoencoder tests.
- 370 timeouts total across the 5 cancelled shards.
- Baseline cancellation rate: 6 of the 20 test shards hit the 45-minute wall clock even before memory issues were masking downstream failures.

## Hypothesis to check (not claims — I can't see the engine source from where I'm working)

Would appreciate pointers or a quick check on whether:

1. `TensorMatMul` on CPU dispatches through a native BLAS (AiDotNet.Native.OpenBLAS is referenced as a package), or falls back to a managed triple-nested loop for specific tensor ranks / element types. The library tagline says "hand-tuned AVX2/FMA SIMD kernels" which is promising — but it's worth confirming the paths exercised under a standard `dotnet test` environment with no special env flags actually hit those kernels rather than a scalar fallback.
2. `Conv2D` on CPU uses im2col + GEMM, winograd, or a direct loop. The 224×224 × 3×3 × 64–512-channel stack in VGG is ~billions of FMAs per forward; a direct-loop implementation without SIMD would alone account for the observed wall-clock.
3. `ScaledDotProductAttention` on CPU has a fused implementation (single-pass over K,V with online softmax) or materializes the full `[N,N]` attention matrix. DiT at patch resolution has `N=256`; full materialization with softmax → matmul is maybe 2× slower than Flash-style fused but still tolerable. Worth confirming.
4. Whether the default CPU engine uses multi-threading for large matmuls (via `Parallel.For` or ThreadPool partitioning). The Windows runners have 2 cores; the Linux runners have 4. Test wall clock should scale inverse-linearly with cores for large GEMMs if threading is on — if it doesn't, that's a clear gap.

## What I'm NOT asking for

- No specific fix, no code changes on this side.
- Not claiming any of the above is slow — I can't see the source. Just documenting what the downstream workload is doing so the maintainer side has visibility when prioritizing.
- Not a bug report — the library is working correctly, just potentially leaving perf on the table for this particular large-model-on-CPU workload.

## Downstream links

- [AiDotNet#1136](https://github.com/ooples/AiDotNet/issues/1136) — the overarching "unblock cancelled CI shards" effort
- [AiDotNet#1137](https://github.com/ooples/AiDotNet/pull/1137) — noise-predictor lazy init
- [AiDotNet#1140](https://github.com/ooples/AiDotNet/pull/1140) — MHA + SelfAttention lazy init + Dispose→pool cascade
- [AiDotNet#1141](https://github.com/ooples/AiDotNet/pull/1141) — scalar NumOps loops → Engine ops in hot-path layers

Closing this repo's loop on the AiDotNet side is in progress. Flagging the engine-side observations here so there's a paper trail if/when the test-harness gets GPU runners or the engine path gets profiled for this specific model shape.

**Comment by ooples** [comment]
Thanks for the detailed writeup — this is very useful for the ongoing `feat/finish-mkl-replacement` branch. Went through your four hypotheses against the current master (`72b01c1`); results below. **The headline is #3: you're right to suspect SDPA, and it's worse than you thought.**

## 1. `TensorMatMul` dispatch on CPU

**Confirmed: hits optimized path for float/double, not a scalar fallback.**

Dispatch chain for 2D float matmul:
1. `TensorMatMul` → `TensorMatMul2D` (`CpuEngine.cs:8496`)
2. `MatrixMultiplyHelper.TryGemm` (threshold `m*k*n ≥ 4096`, so most real matmuls qualify)
3. `BlasProvider.TryGemm`: tries MKL.NET first, then native `cblas_sgemm` (OpenBLAS), else returns false
4. If BLAS unavailable → `SimdGemm.Sgemm` (our hand-tuned AVX2 tiled kernel, `Engines/Simd/SimdGemm.cs`)
5. Scalar parallel `Parallel.For` triple-loop only for non-float/non-double `T`

Under a standard `dotnet test` on Windows CI, the MKL.NET NuGet (`MKL.NET.win-x64`) gets loaded from the runtime directory on first use — should be hitting MKL SGEMM. On Linux CI, it depends on whether `AiDotNet.Native.OpenBLAS` is present; if not, falls back to `SimdGemm.Sgemm` which is still AVX2-vectorized. Can be verified at runtime via `BlasProvider.BackendName` — we could add a log line to the test harness to confirm.

## 2. `Conv2D` on CPU

**Confirmed: im2col+GEMM, Winograd, direct SIMD — not a managed triple-loop.**

Five-layer dispatch in `CpuEngine.cs:6387` (Conv2DWithIm2ColFloat):
1. oneDNN (`TryConv2DOneDnn`) — if `AiDotNet.Native.OneDNN` package is present. This would be the fastest path.
2. `FusedConvHelper.Conv2DFused` — im2col + GEMM with cache tiling, uses BLAS underneath
3. `WinogradHelper.Conv2DWinograd` — for 3×3 stride=1 convolutions (the exact VGG inner-block case)
4. `SimdConvHelper.CanUseSimdConv` — direct AVX2 convolution for small 3×3 stride=1
5. Generic `Conv2DWithIm2ColGemm` — im2col + BLAS GEMM fallback

For VGG16BN 224×224 × 3×3 × 64–512 channels (stride=1): **path 3 (Winograd)** is the one taken, which is typically 2-3× faster than im2col+GEMM for exactly this shape. Only falls through to im2col+GEMM if Winograd's thresholds don't match.

double precision path (6235) mirrors this but with double-precision helpers.

Not a concern — VGG's wall-clock shouldn't be dominated by direct-loop convolution.

## 3. `ScaledDotProductAttention` — **confirmed problem**

**You're right: the default CPU path is a scalar virtual-dispatch triple-loop, no BLAS, no SIMD.**

`CpuEngine.ScaledDotProductAttention<T>` at `CpuEngine.cs:14696` does:

```csharp
Parallel.For(0, batch * heads, bh =>
{
    for (int i = 0; i < seqQ; i++)
    for (int j = 0; j < seqK; j++)
    {
        T sum = numOps.Zero;
        for (int k = 0; k < d_k; k++)
            sum = numOps.Add(sum, numOps.Multiply(queryData[...], keyData[...]));
        // ...
    }
});
```

`numOps.Multiply`/`Add` are interface calls through `INumericOperations<T>` — one virtual dispatch per FMA. At ~20 ns/call, and DiT-XL's 75M FMAs per SDPA call × 28 blocks, that's on the order of **40 seconds per forward pass spent in SDPA alone**. For N=256 that's ~2000× slower than a BLAS matmul of the same FMA count.

**Fix: `TensorScaledDotProductAttention` already exists at `CpuEngine.cs:26582`** and does the right thing:
```csharp
var scores = TensorMatMul(query, keyT);
var scaledScores = TensorMultiplyScalar(scores, scale);
var attnWeights = TensorSoftmax(scaledScores, axis);
var output = TensorMatMul(attnWeights, value);
```

Three options, in order of effort:
1. **Quick fix**: rewrite `ScaledDotProductAttention` (the slow one) to delegate to `TensorScaledDotProductAttention` + re-derive the mask/scale/weights-out behavior. Preserves existing API. 1-day work.
2. **Medium**: wire up existing `FlashAttention` (`CpuEngine.cs:14985`) — a tiled fused implementation that exists but isn't on the default path. Better asymptotic memory for long N, but for DiT N=256 the win is mostly eliminating the virtual dispatch.
3. **Long-term**: route all AiDotNet consumers through `TensorScaledDotProductAttention` (shorter name + no mask arg) and deprecate the scalar one.

I'd recommend (1) as a separate quick PR targeting this repo. It's independent of the ongoing MKL replacement work and directly addresses the DiT-XL wall-clock gap.

## 4. Multi-threading for large matmul

**Confirmed: Parallel.For triggered above `ParallelWorkThreshold = 20M` ops.**

`SimdGemm.SgemmTiled` (`Engines/Simd/SimdGemm.cs:236`) dispatches:
- Sequential `SgemmTiled` for `m*k*n < 20M`
- 1D parallel (`SgemmTiledParallelM`) when row blocks saturate cores
- 2D parallel (`SgemmTiledParallel2D`, row × col grid) when one dimension doesn't saturate

Thread count = `CpuParallelSettings.MaxDegreeOfParallelism`, defaulting to `Environment.ProcessorCount` (logical cores — iter 5 added SMT oversubscription since GEMM is load-port-limited on Zen).

For 2-core Windows CI: `maxThreads=2`, `numRowBlocks=9 @ M=1024`, 2D grid → 1D parallel, 9 work units for 2 workers. Should scale ~1.8×. Large GEMMs like DiT-XL `[1024,1152]×[1152,4608]` (5.4 GFLOP) are well above threshold.

For 4-core Linux: `maxThreads=4`, same logic, expected ~3.5× scaling.

MKL separately manages its own threadpool, and on CI it may default to `Environment.ProcessorCount` as well. MKL and our Parallel.For don't coordinate, which can cause over-subscription — this is known but hasn't been measured to regress in practice.

**One concern**: the 20M threshold may be wrong for 2-core CI. It was tuned on a 16-core Ryzen where thread-pool overhead dominates at small sizes. On 2 cores the overhead should be ~10× lower, so we might benefit from a lower threshold there (dispatching parallel at say 4M). Untested.

## Current related work

`feat/finish-mkl-replacement` branch is actively working on getting our SIMD GEMM past MKL at DiT-XL shapes. Phase 0 baseline (docs/mkl-replacement/baseline/baseline-iter17.md) confirms our current gap is 1.31× at Square 1152², 1.52× at Square 4608², 1.37-1.45× at DiT per-block. Iter 18a landed a masked AVX2 edge kernel that closed per-head attention A·V from 10.3× slower to 2.7× slower (3.48× speedup on that shape). Iter 19 (oneDNN-style prefetcht2 on next A panel) is currently running its baseline benchmark.

Once that branch beats MKL at every shape, we flip `_useMklNet=false` as default and the MKL.NET package reference goes away. That addresses part of your concern but **does not address the SDPA scalar-loop issue, which is a separate quick win that probably saves more wall-clock on DiT-XL than the entire MKL replacement effort**. Worth filing as its own issue.

Happy to file the SDPA fix as a separate PR on this branch if you want — it's 20 lines plus tests.

**Comment by ooples** [comment]
Update: shipped the three code fixes from my earlier comment on the `feat/finish-mkl-replacement` branch. Landing commits listed below. Runs are on the 16-core dev box; CI wall-clock impact will show up on the 2-4 core runners.

## Landed

**cd954be — SDPA BLAS-backed float fast path**

`ScaledDotProductAttention<T>` at `CpuEngine.cs:14696` now dispatches to two batched SGEMMs per head (Q·K^T via `BlasProvider.TryGemmEx(transB=true)` then P·V), parallel over `batch*heads`, for `T=float`. Non-float `T` continues on the original scalar path — no behavior change for those callers.

Direct before/after on Ryzen 9 3950X at DiT-XL attention shape `[B=4, H=16, S=256, d=72]`:
- Float BLAS fast path: **25.3 ms**
- Double scalar legacy path: **93.3 ms**
- Speedup: **3.68×** (conservative — double-vs-float on AVX2 is about 2×, so pure-float scalar-vs-BLAS is wider)

DiT-XL forward-pass impact: 28 SDPA calls × 25 ms = 0.7 s, down from ~2.6 s with the scalar path. Saves ~2 s per forward. On 2-4 core CI runners absolute savings will be smaller but the ratio holds — scalar SDPA is bound by per-op virtual dispatch overhead, which parallelizes no differently than the BLAS path.

Correctness: 39/39 attention tests pass on net10.0.

**14efda4 — ParallelWorkThreshold scales with `Environment.ProcessorCount`**

Hypothesis #4 from your issue: iter 2's 20M threshold was tuned on 16-core Ryzen and leaves 2-4 core CI runners sequential on medium matmuls. Change:

```
 2 cores:  2M   (CI Windows — parallel at 128² and above)
 4 cores:  5M   (CI Linux)
 8 cores: 10M
16 cores: 20M   (iter 2 default — unchanged)
```

Preserves exact iter-2 behavior at 16+ cores. Low-core boxes now get SgemmTiledParallelM/2D kicking in at smaller problem sizes. Can't benchmark this one on the dev box (threshold unchanged at 16 cores), so the impact will land in your CI metrics directly.

Correctness: 118/118 matmul + GEMM + DeterministicMode + SDPA + Attention tests pass on net10.0.

## Verified, no change needed

- **`TensorMatMul`** (hypothesis #1): 4-tier dispatch already (NativeInferencePool → raw calli → native delegate → MKL-verified → `SimdGemm` SIMD fallback). Float/double hit BLAS. Only non-float/double `T` goes scalar.
- **`Conv2D`** (hypothesis #2): 5-tier dispatch (oneDNN → FusedConv im2col+BLAS → Winograd → SIMD direct → im2col+BLAS fallback). VGG16BN 3×3-stride=1 hits Winograd. No scalar.
- **`FusedLinear`**: Same 4-tier BLAS dispatch as TensorMatMul for float/double, plus pinned-weights `NativeInferencePool` path for inference. `CpuEngine.cs:22360+`.
- **`LayerNorm`** float path: direct scalar-float loop (not virtual dispatch), parallelized at `batchSize * features ≥ 50K`. `CpuEngine.cs:14112+`. RyuJIT autovectorizes the reduction. Non-float `T` goes through `INumericOperations` virtual dispatch but that's outside DiT-XL's path.

## NOT addressed in this PR (separate follow-ups if they matter)

- **`FlashAttention<T>`** at `CpuEngine.cs:14985` — also a scalar virtual-dispatch loop. Not on the default path for DiT-XL (which uses SDPA), so kept deferred. If downstream ever opts in, same fix pattern as the SDPA fast path would apply.
- **`Conv2DBackwardKernel`** scalar-T path — same story, only relevant when `T` is not float/double.
- **oneDNN availability** — if `AiDotNet.Native.OneDNN` isn't installed on CI, Conv2D falls through to FusedConv im2col+BLAS (still fast, but one rung down). Worth checking in your environment probe.

## Ongoing MKL replacement work

The `feat/finish-mkl-replacement` branch is still closing the square-shape GEMM gap to MKL (currently 1.31× at 1152², 1.52× at 4608² per iter 17 baseline). Iters 18a (masked edge kernel → A·V 3.48× speedup) and 19 (oneDNN-style next-panel prefetch) landed. Next are iters 20-23 (K-unroll, per-µarch tiles, NT stores for large C, libxsmm-style JIT). The SDPA fix here is orthogonal to that work — even at MKL-speed GEMM, the scalar SDPA would have been the dominant cost.

**Comment by ooples** [comment]
## Recommendation: close as substantially addressed

Since this issue was filed, the following work has landed or is in active PR review targeting all four bottleneck areas identified:

| Bottleneck | What shipped | PR/Issue |
|---|---|---|
| `TensorMatMul` / `FusedLinear` | MKL.NET removal → SimdGemm AVX2 blocked kernel (at-or-faster than MKL on all DiT-XL shapes) | #163 (merged) |
| `Conv2D` / `Conv2DBackward` | SimdGemm for Conv im2col+GEMM path + `FusedConv2DBiasActivationOp` JIT fusion | #163 (merged) + #180 (PR #187) |
| `ScaledDotProductAttention` | SDPA BLAS-backed float fast path (commit cd954be, landed on main via #163) | #163 (merged) |
| `LayerNorm` / `GroupNorm` | `GroupNormOp` IR op + `FusedGroupNormActivationOp` (GroupNorm+SiLU single-pass kernel, 80 fusions per SD15 forward) | #178, #179 (PR #187) |

Additionally:
- Deterministic-by-default (#174) ensures the SimdGemm bit-exact path is the default — no MKL multi-threaded GEMM overhead
- JIT fusion patterns 11-14 (#181) eliminate ~160 intermediate tensor materializations per SD15 forward pass
- The SDPA fix in the owner's comment (cd954be) directly addressed the worst finding ("worse than you thought")

**Suggestion:** close this issue and re-open if CI benchmarks still show timeouts after PRs #183, #185, #186, #187 merge. The remaining perf ceiling is likely the memory planning pass (#182), which is being worked on next.
